src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDD.cs

```diff
@@ -392,13 +377,17 @@ static BDD DeserializeWithFactory(long[] arcs, Func<int, BDD?, BDD?, BDD> create
                         int zeroId = (int)((arc >> zero_node_shift) & node_mask);
 
                         // the BDD nodes for the children are guaranteed to exist already due to the topological order
-                        nodes[i] = createBDD(ord, nodes[oneId], nodes[zeroId]);
+                        nodes[i] = algebra.GetOrCreateBDD(ord, nodes[oneId], nodes[zeroId]);
                     }
                 }
 
                 //the result is the final BDD in the nodes array
                 return nodes[k - 1];
             }
+            finally
+            {
+                algebra.Lock.ExitWriteLock();
+            }
```

Comment:
What is it we're protecting access to here with the GetOrCreateBDD call?  Is it just the state here, or are we trying to synchronize access across multiple calls, or something else?

I ask because it seems like the `HashSet<BDD>` could be replaced by a `ConcurrentDictionary<BDD, BDD>`, and GetOrCreateBDD could become:
```C#
var key = new BDD(ordinal, one, zero);
return _bddCache.GetOrAdd(key, key);
```

Reply:
Thinking about this more I do think ConcurrentDictionary might be the better approach. The main concern is indeed the synchronization of that datastructure, ensuring a unique representation for each BDD. I'll change this.